devel/helloSlider.lua

```diff
@@ -0,0 +1,212 @@
+local SLIDER_RIGHT_END = to_pen{ch=181, fg=COLOR_GREY, bg=COLOR_BLACK}
+local SLIDER_TAB_LEFT = to_pen{ch=60, fg=COLOR_BLACK, bg=COLOR_YELLOW}
+local SLIDER_TAB_CENTER = to_pen{ch=9, fg=COLOR_BLACK, bg=COLOR_YELLOW}
+local SLIDER_TAB_RIGHT = to_pen{ch=62, fg=COLOR_BLACK, bg=COLOR_YELLOW}
+
+function Slider:onRenderBody(dc, rect)
+    local left_idx = self.get_idx_fn()
+    local width_per_idx = Slider_get_width_per_idx(self)
+    -- draw track
+    dc:seek(1,0)
+    dc:char(nil, SLIDER_LEFT_END)
+    dc:char(nil, SLIDER_TRACK)
+    for stop_idx=1,self.num_stops-1 do
+        local track_stop_pen = SLIDER_TRACK_STOP_SELECTED
+        local track_pen = SLIDER_TRACK_SELECTED
+        if left_idx ~= stop_idx then
+            track_stop_pen = SLIDER_TRACK_STOP
+            track_pen = SLIDER_TRACK
+        elseif left_idx == stop_idx then
+            track_pen = SLIDER_TRACK
+        end
+        dc:char(nil, track_stop_pen)
+        for i=2,width_per_idx do
+            dc:char(nil, track_pen)
+        end
+    end
+    if left_idx >= self.num_stops then
+        dc:char(nil, SLIDER_TRACK_STOP_SELECTED)
+    else
+        dc:char(nil, SLIDER_TRACK_STOP)
+    end
+    dc:char(nil, SLIDER_TRACK)
+    dc:char(nil, SLIDER_RIGHT_END)
+    -- draw tab
+    dc:seek(width_per_idx*(left_idx-1)+2)
+    dc:char(nil, SLIDER_TAB_LEFT)
+    dc:char(nil, SLIDER_TAB_CENTER)
+    dc:char(nil, SLIDER_TAB_RIGHT)
+    -- manage dragging
+    if self.is_dragging_target then
+        Slider_do_drag(self, width_per_idx)
+    end
+    if df.global.enabler.mouse_lbut_down == 0 then
+        self.is_dragging_target = nil
+        self.is_dragging_idx = nil
+    end
+end
+
+
+
+
+
+
+
+
+
+
+
+local gui = require('gui')
+local widgets = require('gui.widgets')
+
+--
+-- RangerWindow
+--
+
+RangerWindow = defclass(RangerWindow, widgets.Window)
+RangerWindow.ATTRS {
+    frame_title='Hello, Slider!',
+    frame={w=25, h=8},
+    resizable=true,
+    resize_min={w=25, h=8},
+}
+
+function RangerWindow:init()
+    local LEVEL_OPTIONS = {
+        {label='Low', value=1},
+        {label='Medium', value=2},
+        {label='High', value=3},
+        {label='Pro', value=4},
+        {label='Insane', value=5},
+    }
+
+    self:addviews{
+        widgets.CycleHotkeyLabel{
+            view_id='level',
+            frame={l=1, t=0, w=16},
+            label='Level:',
+            label_below=true,
+            key_back='CUSTOM_SHIFT_C',
+            key='CUSTOM_SHIFT_V',
+            options=LEVEL_OPTIONS,
+            initial_option=LEVEL_OPTIONS[1].value,
+            on_change=function(val)
+                self.subviews.level:setOption(val)
+            end,
+        },
+        Slider{
+            frame={l=1, t=3},
+            num_stops=#LEVEL_OPTIONS,
+            get_idx_fn=function()
+                return self.subviews.level:getOptionValue()
+            end,
+            on_change=function(idx) self.subviews.level:setOption(idx) end,
+        },
+    }
```

Comment:
this code should be added to https://github.com/DFHack/scripts/blob/master/devel/hello-world.lua, with frame sizes adjusted to accommodate.